### ISSUE TITLE
Jax Neural Network Regression Always Predicts the Mean

### ISSUE THREAD

**Issue by LenCewa** [google/jax#2605]
Hello,
I implemented a small NN in Jax to predict simple functions, like polynomials or trigonometric functions. I took the inspiration for the design of the NN from the [colab](https://colab.research.google.com/github/google/jax/blob/master/docs/notebooks/neural_network_with_tfds_data.ipynb). However, my NN always predicts the average of the `train_labels` (see figure below the code snippet). I have a quite big NN - `[1, 17, 17, 1]` - to learn e.g. `f(x) = x^2`. I don't have much experience with NN but I heard of the universial function approximation theorem and therefore thought this regression problem shouldn't be a big deal. I wasn't sure if this is the right format where one would post this question - but as it's related to jax I thought I give it a try (instead of using stackoverflow).

- Network architecture is [1, 17, 17, 1] (different architectures with the same amount of parameter failed either)

- Activation = `ReLu` (but others didn't work either)

My code:

```
def random_layer_params(m, n, key, scale=1e-2):
  w_key, b_key = random.split(key)
  return scale * random.normal(w_key, (n, m)), scale * random.normal(b_key, (n,))

def init_network_params(sizes, key):
  keys = random.split(key, len(sizes))
  return [random_layer_params(m, n, k) for m, n, k in zip(sizes[:-1], sizes[1:], keys)]

layer_sizes = [1, 17, 17, 1]
params = init_network_params(layer_sizes, random.PRNGKey(0))
initial_params = params
step_size = 0.001

training_generator = jnp.linspace(0, 2*jnp.pi, num=50)
train_values = jnp.array(training_generator)
train_labels = [x**2 for x in train_values]
train_labels = jnp.array(train_labels)
train_values = jnp.array(training_generator).reshape(-1,1)

test_generator = jnp.linspace(0, 2*jnp.pi, num=500)
test_values = jnp.array(test_generator)
test_labels = [x**2 for x in test_values]
test_labels = jnp.array(test_labels)
test_values = jnp.array(test_generator).reshape(-1,1)

def relu(x):
    return jnp.maximum(0, x)

def predict(params, inputs):
    activations = inputs
    for w, b in params[:-1]:
        outputs = jnp.dot(w, activations) + b
        activations = relu(outputs)

    final_w, final_b = params[-1]
    result = jnp.dot(final_w, activations) + final_b
    return result

batched_predict = vmap(predict, in_axes=(None, 0))

def loss(params, train_values, train_labels):
    preds = batched_predict(params, train_values)[:, 0]
    return jnp.sum(jnp.square(preds - train_labels))

#@jit
def update(params, x, y):
    grads = grad(loss)(params, x, y)
    return [(w - step_size * dw, b - step_size * db) for (w, b), (dw, db) in zip(params, grads)]

for i in range(10000):
    params = update(params, train_values, train_labels)
```
![Figure_1](https://user-images.githubusercontent.com/11056759/78460687-36e68a00-76c3-11ea-930f-802bbcc8b737.png)

**Comment by mattjj** [comment]
Thanks for the question!

I think perhaps you just need some step size tuning. Changing `step_size = 1e-4`, and setting up some loss function printing like this:

```python
for i in range(10000):
    params = update(params, train_values, train_labels)
    if i % 100 == 0:
        print(i, loss(params, train_values, train_labels))
```

I get this:

```
0 15893.639
100 3211.0073
200 1328.9572
300 222.83786
400 3887.205
500 1600.1885
...
9000 25.13343
9100 24.814024
9200 24.422073
9300 22.984049
9400 21.983908
9500 20.755407
9600 17.883072
9700 16.26672
9800 16.372345
9900 18.52781
```

(I put back the `@jit` decorator too.)

So the loss is definitely going down! A good rule of thumb is to make sure that for small enough step size the loss is going down pretty steadily, then try growing the step size from there.

Here's what the plot looks like:

```python
import matplotlib.pyplot as plt
plt.plot(train_labels, label='labels')
plt.plot(batched_predict(params, train_values).ravel(), label='prediction')
plt.savefig('plot.png')
```

![plot](https://user-images.githubusercontent.com/1458824/78462766-09311f00-768a-11ea-96af-ef43ddd0872b.png)

WDYT?

**Comment by LenCewa** [comment]
Thanks for the quick response! It helped to solve my problems - and yes I also put the `@jit` decorator back too :)
